### ISSUE TITLE
setVolume before loading audio doesn't change volume

### ISSUE THREAD

**Issue by YukunXia** [ryanheise/just_audio#439]
**Which API doesn't behave as documented, and how does it misbehave?**
player.setVolume doesn't change the volume if no audio source has been loaded yet.

**Minimal reproduction project**
Link: https://github.com/YukunXia/just_audio

**To Reproduce (i.e. user steps, not code)**
1, Go to just_audio/example
2, `flutter run -d macos`
3, Click the volume button, change the value to 0 or 0.1
4, Click the `load audio` button
5, Click the play button

**Error messages**

No error message

**Expected behavior**

the audio is played in full volume, rather than 0 or 0.1 as just set.

**Screenshots**

**Desktop (please complete the following information):**
 - macOS
 - Version: either the one in the repo or ^0.8.0

**Flutter SDK version**
```
flutter doctor                   
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, 2.3.0-24.1.pre, on macOS 11.1 20C69 darwin-x64, locale en-US)
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more
      details.
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.2)
[✓] VS Code (version 1.57.1)
[✓] Connected device (2 available)
```

**Additional context**

**Comment by ryanheise** [comment]
> **Expected behavior**
> 
> the audio is played in full volume, rather than 0 or 0.1 as just set.

Is this expectation right? Why wouldn't you expect the volume to be played the way it was set?

**Comment by YukunXia** [comment]
I'd imagine that once a player object is created, if I change the volume at the initial stage, that change should apply to later audio playing. Loading audio should be one step of the initialization, in parallel with other settings, eg, the default volume. 

Anyway, following the reproduction steps I mentioned, the volume change has been verified in the volume stream and visualized in the slider, but actually, there's no change in the player. It's a little bit confusing.

**Comment by ryanheise** [comment]
I haven't had a chance to investigate this yet (and I will), but I was just trying to understand your expected behaviour section as it seemed you had it around the opposite way.

To me, the expected behaviour is not "the audio is played in full volume, rather than 0 or 0.1 as just set".

Instead, the expected behaviour is "the audio is played at 0 or 0.1 as just set, rather than in full volume".

**Comment by YukunXia** [comment]
Maybe I didn't express clearly, but I'd expect the audio to be played at 0 or 0.1 too...

**Comment by schisms** [comment]
I have the same issue, but honestly was too lazy to investigate and make an issue of it.

A temporary, lazy-man solution is:
`await _player.setVolume(_player.volume);`

**Comment by YukunXia** [comment]
> I have the same issue, but honestly was too lazy to investigate and make an issue of it.
> 
> A temporary, lazy-man solution is:
> `await _player.setVolume(_player.volume);`
> 
> You have to call setVolume once player has an AudioSource loaded.

Hi @schisms , where do you put that line in? I put that into the initState function, but it makes no change...

**Comment by YukunXia** [comment]
> I have the same issue, but honestly was too lazy to investigate and make an issue of it.
> 
> A temporary, lazy-man solution is:
> `await _player.setVolume(_player.volume);`
> 
> You have to call setVolume once player has an AudioSource loaded.

@schisms  Thank you! This trick works. But it seems to cause another minor issue. Every time I open the volume slider, the volume will appear as full for a short moment, and then show the correct volume. Did you also observe that?

**Comment by ryanheise** [comment]
Hi @YukunXia I'm getting a 404 on your reproduction project.

**Comment by YukunXia** [comment]
Oops, I thought you have seen that. Let me just copy the `main.dart` here. There's no big change. I just created a button for audio loading, and deleted the audio loading at the init stage.

```dart
// This is a minimal example demonstrating a play/pause button and a seek bar.
// More advanced examples demonstrating other features can be found in the same
// directory as this example in the GitHub repository.

import 'package:audio_session/audio_session.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:just_audio/just_audio.dart';
import 'package:just_audio_example/common.dart';
import 'package:rxdart/rxdart.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final AudioPlayer _player = AudioPlayer();

  @override
  void initState() {
    super.initState();
    SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle(
      statusBarColor: Colors.black,
    ));
    _init();
  }

  Future<void> _init() async {
    // Inform the operating system of our app's audio attributes etc.
    // We pick a reasonable default for an app that plays speech.
    final session = await AudioSession.instance;
    await session.configure(AudioSessionConfiguration.speech());
    // Listen to errors during playback.
    _player.playbackEventStream.listen((event) {},
        onError: (Object e, StackTrace stackTrace) {
      print('A stream error occurred: $e');
    });
  }

  @override
  void dispose() {
    // Release decoders and buffers back to the operating system making them
    // available for other apps to use.
    _player.dispose();
    super.dispose();
  }

  /// Collects the data useful for displaying in a seek bar, using a handy
  /// feature of rx_dart to combine the 3 streams of interest into one.
  Stream<PositionData> get _positionDataStream =>
      Rx.combineLatest3<Duration, Duration, Duration?, PositionData>(
          _player.positionStream,
          _player.bufferedPositionStream,
          _player.durationStream,
          (position, bufferedPosition, duration) => PositionData(
              position, bufferedPosition, duration ?? Duration.zero));

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: SafeArea(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              TextButton(
                onPressed: () async {
                  // Try to load audio from a source and catch any errors.
                  try {
                    await _player.setAudioSource(AudioSource.uri(Uri.parse(
                        "https://s3.amazonaws.com/scifri-episodes/scifri20181123-episode.mp3")));
                  } catch (e) {
                    print("Error loading audio source: $e");
                  }
                },
                child: Text("load audio"),
              ),
              // Display play/pause button and volume/speed sliders.
              ControlButtons(_player),
              // Display seek bar. Using StreamBuilder, this widget rebuilds
              // each time the position, buffered position or duration changes.
              StreamBuilder<PositionData>(
                stream: _positionDataStream,
                builder: (context, snapshot) {
                  final positionData = snapshot.data;
                  return SeekBar(
                    duration: positionData?.duration ?? Duration.zero,
                    position: positionData?.position ?? Duration.zero,
                    bufferedPosition:
                        positionData?.bufferedPosition ?? Duration.zero,
                    onChangeEnd: _player.seek,
                  );
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}

/// Displays the play/pause button and volume/speed sliders.
class ControlButtons extends StatelessWidget {
  final AudioPlayer player;

  ControlButtons(this.player);

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        // Opens volume slider dialog
        IconButton(
          icon: Icon(Icons.volume_up),
          onPressed: () {
            showSliderDialog(
              context: context,
              title: "Adjust volume",
              divisions: 10,
              min: 0.0,
              max: 1.0,
              stream: player.volumeStream,
              onChanged: player.setVolume,
            );
          },
        ),

        /// This StreamBuilder rebuilds whenever the player state changes, which
        /// includes the playing/paused state and also the
        /// loading/buffering/ready state. Depending on the state we show the
        /// appropriate button or loading indicator.
        StreamBuilder<PlayerState>(
          stream: player.playerStateStream,
          builder: (context, snapshot) {
            final playerState = snapshot.data;
            final processingState = playerState?.processingState;
            final playing = playerState?.playing;
            if (processingState == ProcessingState.loading ||
                processingState == ProcessingState.buffering) {
              return Container(
                margin: EdgeInsets.all(8.0),
                width: 64.0,
                height: 64.0,
                child: CircularProgressIndicator(),
              );
            } else if (playing != true) {
              return IconButton(
                icon: Icon(Icons.play_arrow),
                iconSize: 64.0,
                onPressed: player.play,
              );
            } else if (processingState != ProcessingState.completed) {
              return IconButton(
                icon: Icon(Icons.pause),
                iconSize: 64.0,
                onPressed: player.pause,
              );
            } else {
              return IconButton(
                icon: Icon(Icons.replay),
                iconSize: 64.0,
                onPressed: () => player.seek(Duration.zero),
              );
            }
          },
        ),
        // Opens speed slider dialog
        StreamBuilder<double>(
          stream: player.speedStream,
          builder: (context, snapshot) => IconButton(
            icon: Text("${snapshot.data?.toStringAsFixed(1)}x",
                style: TextStyle(fontWeight: FontWeight.bold)),
            onPressed: () {
              showSliderDialog(
                context: context,
                title: "Adjust speed",
                divisions: 10,
                min: 0.5,
                max: 1.5,
                stream: player.speedStream,
                onChanged: player.setSpeed,
              );
            },
          ),
        ),
      ],
    );
  }
}
```

**Comment by ryanheise** [comment]
This should be fixed in 0.9.1, let me know if you still encounter the issue otherwise I'll close it.

> Every time I open the volume slider, the volume will appear as full for a short moment, and then show the correct volume. Did you also observe that?

I observe that regardless of this fix, it is more related to a the way the UI in the example is written:

```dart
              Slider(
                divisions: divisions,
                min: min,
                max: max,
                value: snapshot.data ?? 1.0,
                onChanged: onChanged,
              ),
```

Before the snapshot has loaded, it defaults to 1.0. You could change the default to the actual value of `player.volume` and it will fix that glitch.
